src/librustc_mir/transform/unreachable_prop.rs

```diff
@@ -0,0 +1,97 @@
+//! A pass that propagates the unreachable terminator of a block to its predecessors
+//! when all of their successors are unreachable. This is achieved through a
+//! post-order traversal of the blocks.
+
+use rustc::ty::TyCtxt;
+use rustc::mir::*;
+use rustc_data_structures::fx::{FxHashSet, FxHashMap};
+use std::borrow::Cow;
+use crate::transform::{MirPass, MirSource};
+use crate::transform::simplify;
+
+pub struct UnreachablePropagation;
+
+impl MirPass<'_> for UnreachablePropagation {
+    fn run_pass<'tcx>(&self, _tcx: TyCtxt<'tcx>, _: MirSource<'tcx>, body: &mut BodyCache<'tcx>) {
+        let mut unreachable_blocks = FxHashSet::default();
+        let mut replacements = FxHashMap::default();
+
+        for (bb, bb_data) in traversal::postorder(body) {
+            let terminator = bb_data.terminator();
+
+            if let TerminatorKind::Unreachable = terminator.kind {
+                unreachable_blocks.insert(bb);
+            } else {
+                let is_unreachable_pred = |succ: &'_ BasicBlock| unreachable_blocks.contains(succ);
+                let mut mut_terminator = terminator.clone();
+
```

Comment:
Please check for any asm statements in the block's statement list and if there are any, don't attempt the optimization. Add a HACK comment so we find it again until we have a general solution. You can probably do the check by calling the already existing function (I think? cc @Centril ) in the nop match optimization

Reply:
https://github.com/rust-lang/rust/pull/66282/files#diff-c01519ef1a5124af6e510fbe1de26197R167